#### Code by Larissa Evaldt

#### Importing libraries and reading dataset

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('data/garda.csv', index_col=[0])

In [3]:
df.shape

(1476, 1)

#### Dataset has 1476 rows and 1 column. Let's have a look on the first rows:

In [4]:
df.head()

Garda Station
0  63101 Balbriggan, D.M.R. Northern Division
1  63101 Balbriggan, D.M.R. Northern Division
2  63101 Balbriggan, D.M.R. Northern Division
3  63101 Balbriggan, D.M.R. Northern Division
4  63101 Balbriggan, D.M.R. Northern Division

#### We have a lot of duplicated entries, because this dataset was dividided by types of crimes, it had a row for each type of crime and the number of times that the type of crime happened in each station. Let's get only the unique entries:

In [5]:
df['Garda Station'].unique()

array(['63101 Balbriggan, D.M.R. Northern Division',
       '66201 Ballyfermot, D.M.R. Western Division',
       '63201 Ballymun, D.M.R. Northern Division',
       '65101 Blackrock, Co Dublin, D.M.R. Eastern Division',
       '66101 Blanchardstown, D.M.R. Western Division',
       '62101 Bridewell Dublin, D.M.R. North Central Division',
       '65201 Cabinteely, D.M.R. Eastern Division',
       '66102 Cabra, D.M.R. Western Division',
       '66202 Clondalkin, D.M.R. Western Division',
       '63401 Clontarf, D.M.R. Northern Division',
       '63301 Coolock, D.M.R. Northern Division',
       '64101 Crumlin, D.M.R. Southern Division',
       '61101 Donnybrook, D.M.R. South Central Division',
       '63202 Dublin Airport, D.M.R. Northern Division',
       '65203 Dun Laoghaire, D.M.R. Eastern Division',
       '65102 Dundrum, D.M.R. Eastern Division',
       '66103 Finglas, D.M.R. Western Division',
       '62202 Fitzgibbon Street, D.M.R. North Central Division',
       '63102 Garristown, 

In [6]:
len(df['Garda Station'].unique())

41

#### We have data for 41 garda stations. Let's drop duplicates and keep only the unique

In [7]:
df = df.drop_duplicates()

In [8]:
df

Garda Station
0           63101 Balbriggan, D.M.R. Northern Division
12          66201 Ballyfermot, D.M.R. Western Division
24            63201 Ballymun, D.M.R. Northern Division
36   65101 Blackrock, Co Dublin, D.M.R. Eastern Div...
48       66101 Blanchardstown, D.M.R. Western Division
60   62101 Bridewell Dublin, D.M.R. North Central D...
72           65201 Cabinteely, D.M.R. Eastern Division
84                66102 Cabra, D.M.R. Western Division
96           66202 Clondalkin, D.M.R. Western Division
108           63401 Clontarf, D.M.R. Northern Division
120            63301 Coolock, D.M.R. Northern Division
132            64101 Crumlin, D.M.R. Southern Division
144    61101 Donnybrook, D.M.R. South Central Division
156     63202 Dublin Airport, D.M.R. Northern Division
168       65203 Dun Laoghaire, D.M.R. Eastern Division
180             65102 Dundrum, D.M.R. Eastern Division
192             66103 Finglas, D.M.R. Western Division
204  62202 Fitzgibbon Street, D.M.R. North Central ...
216         63102 Garristown, D.M.R. Northern Division
228              63402 Howth, D.M.R. Northern Division
240     61102 Irishtown, D.M.R. South Central Division
252  61301 Kevin Street, D.M.R. South Central Division
264    61302 Kilmainham, D.M.R. South Central Division
276               66301 Lucan, D.M.R. Western Division
288               63103 Lusk, D.M.R. Northern Division
300           63302 Malahide, D.M.R. Northern Division
312      62203 Mountjoy, D.M.R. North Central Division
324  61202 Pearse Street, D.M.R. South Central Divi...
336             63403 Raheny, D.M.R. Northern Division
348           66203 Rathcoole, D.M.R. Western Division
360        64201 Rathfarnham, D.M.R. Southern Division
372          64301 Rathmines, D.M.R. Southern Division
384          66302 Ronanstown, D.M.R. Western Division
396             63203 Santry, D.M.R. Northern Division
408            65205 Shankill, D.M.R. Eastern Division
420           63105 Skerries, D.M.R. Northern Division
432  62301 Store Street, D.M.R. North Central Division
444      64102 Sundrive Road, D.M.R. Southern Division
456             63303 Swords, D.M.R. Northern Division
468           64202 Tallaght, D.M.R. Southern Division
480           64302 Terenure, D.M.R. Southern Division

#### Reseting the index

In [9]:
df = df.reset_index(drop=True)

In [10]:
df

Garda Station
0          63101 Balbriggan, D.M.R. Northern Division
1          66201 Ballyfermot, D.M.R. Western Division
2            63201 Ballymun, D.M.R. Northern Division
3   65101 Blackrock, Co Dublin, D.M.R. Eastern Div...
4       66101 Blanchardstown, D.M.R. Western Division
5   62101 Bridewell Dublin, D.M.R. North Central D...
6           65201 Cabinteely, D.M.R. Eastern Division
7                66102 Cabra, D.M.R. Western Division
8           66202 Clondalkin, D.M.R. Western Division
9            63401 Clontarf, D.M.R. Northern Division
10            63301 Coolock, D.M.R. Northern Division
11            64101 Crumlin, D.M.R. Southern Division
12    61101 Donnybrook, D.M.R. South Central Division
13     63202 Dublin Airport, D.M.R. Northern Division
14       65203 Dun Laoghaire, D.M.R. Eastern Division
15             65102 Dundrum, D.M.R. Eastern Division
16             66103 Finglas, D.M.R. Western Division
17  62202 Fitzgibbon Street, D.M.R. North Central ...
18         63102 Garristown, D.M.R. Northern Division
19              63402 Howth, D.M.R. Northern Division
20     61102 Irishtown, D.M.R. South Central Division
21  61301 Kevin Street, D.M.R. South Central Division
22    61302 Kilmainham, D.M.R. South Central Division
23               66301 Lucan, D.M.R. Western Division
24               63103 Lusk, D.M.R. Northern Division
25           63302 Malahide, D.M.R. Northern Division
26      62203 Mountjoy, D.M.R. North Central Division
27  61202 Pearse Street, D.M.R. South Central Divi...
28             63403 Raheny, D.M.R. Northern Division
29           66203 Rathcoole, D.M.R. Western Division
30        64201 Rathfarnham, D.M.R. Southern Division
31          64301 Rathmines, D.M.R. Southern Division
32          66302 Ronanstown, D.M.R. Western Division
33             63203 Santry, D.M.R. Northern Division
34            65205 Shankill, D.M.R. Eastern Division
35           63105 Skerries, D.M.R. Northern Division
36  62301 Store Street, D.M.R. North Central Division
37      64102 Sundrive Road, D.M.R. Southern Division
38             63303 Swords, D.M.R. Northern Division
39           64202 Tallaght, D.M.R. Southern Division
40           64302 Terenure, D.M.R. Southern Division

### We need to get latitude and longitude of these Garda Stations! 

### BUT the Google Maps API is not able to find the addresses in this way that the Irish Censo formats the name of the station. 
### For example, if we search Google Maps for:
* 63101 Balbriggan, D.M.R. Northern Division
#### It will not be able to find. But if instead we search for:
* Balbriggan Garda Station
#### It finds! So let's format our entries to these new format:

#### Looping through every row of the dataset
* For each iteration, select the garda station name. Ex: On the first iteration, original will be = "63101 Balbriggan, D.M.R. Northern Division"
* Then delete the numbers and space from the front. It will become: "Balbriggan, D.M.R. Northern Division"
* Delete everything after the comma. String will be just: "Balbriggan"
* Add " Garda Station" to the end to have the final name, it will be: "Balbriggan Garda Station"
* Update dataframe to new garda name that will be easier for the API to find latitude and longitude.

In [15]:
for x in range(len(df)):
    original = df['Garda Station'][x] #select original name
    without_numbers = original.lstrip('0123456789 ') #delete numbers and space from the front
    without_end = without_numbers.split(',')[0] #delete everything after the comma
    final = without_end + ' Garda Station' #at this point string is smt like for instance: 'Balbriggan Garda Station'
    df['Garda Station'][x] = final #update entry 'x' with new name

In [16]:
df

Garda Station
0          Balbriggan Garda Station
1         Ballyfermot Garda Station
2            Ballymun Garda Station
3           Blackrock Garda Station
4      Blanchardstown Garda Station
5    Bridewell Dublin Garda Station
6          Cabinteely Garda Station
7               Cabra Garda Station
8          Clondalkin Garda Station
9            Clontarf Garda Station
10            Coolock Garda Station
11            Crumlin Garda Station
12         Donnybrook Garda Station
13     Dublin Airport Garda Station
14      Dun Laoghaire Garda Station
15            Dundrum Garda Station
16            Finglas Garda Station
17  Fitzgibbon Street Garda Station
18         Garristown Garda Station
19              Howth Garda Station
20          Irishtown Garda Station
21       Kevin Street Garda Station
22         Kilmainham Garda Station
23              Lucan Garda Station
24               Lusk Garda Station
25           Malahide Garda Station
26           Mountjoy Garda Station
27      Pearse Street Garda Station
28             Raheny Garda Station
29          Rathcoole Garda Station
30        Rathfarnham Garda Station
31          Rathmines Garda Station
32         Ronanstown Garda Station
33             Santry Garda Station
34           Shankill Garda Station
35           Skerries Garda Station
36       Store Street Garda Station
37      Sundrive Road Garda Station
38             Swords Garda Station
39           Tallaght Garda Station
40           Terenure Garda Station

### Getting Latitude and Longitude 

In [17]:
from googlemaps import Client as GoogleMaps
from time import sleep

In [18]:
df['Longitude'] = ""
df['Latitude'] = ""

In [19]:
df.head(1)

Garda Station Longitude Latitude
0  Balbriggan Garda Station

#### Getting latitude and longitude from google maps

In [20]:
gmaps = GoogleMaps('API_KEY_WAS_HERE')

In [21]:
for x in range(len(df)):
    try:
        sleep(1) #to add delay in case of large DFs
        geocode_result = gmaps.geocode(df['Garda Station'][x])        
        df['Latitude'][x] = geocode_result[0]['geometry']['location'] ['lat']
        df['Longitude'][x] = geocode_result[0]['geometry']['location']['lng']
    except IndexError:
        print("Address was wrong...")
    except Exception as e:
        print("Unexpected error occurred.", e )

In [22]:
df

Garda Station Longitude   Latitude
0          Balbriggan Garda Station  -6.19098  53.614378
1         Ballyfermot Garda Station -6.358053  53.344754
2            Ballymun Garda Station -6.263922  53.394374
3           Blackrock Garda Station -6.177503  53.299793
4      Blanchardstown Garda Station -6.380952  53.389924
5    Bridewell Dublin Garda Station -6.274125  53.347081
6          Cabinteely Garda Station -6.150646   53.26079
7               Cabra Garda Station -6.307702  53.365012
8          Clondalkin Garda Station -6.394952  53.323146
9            Clontarf Garda Station  -6.22021  53.363411
10            Coolock Garda Station -6.201027  53.390315
11            Crumlin Garda Station -6.315002   53.31958
12         Donnybrook Garda Station -6.235774   53.32172
13     Dublin Airport Garda Station -6.244229  53.429669
14      Dun Laoghaire Garda Station -6.133399  53.289923
15            Dundrum Garda Station -6.242477  53.289491
16            Finglas Garda Station -6.306434  53.389656
17  Fitzgibbon Street Garda Station -6.255665  53.357832
18         Garristown Garda Station -6.383217  53.567335
19              Howth Garda Station  -6.06862  53.387424
20          Irishtown Garda Station -6.223035  53.338111
21       Kevin Street Garda Station -6.269445  53.338328
22         Kilmainham Garda Station -6.304575  53.341948
23              Lucan Garda Station -6.450605  53.356019
24               Lusk Garda Station -6.167394  53.523377
25           Malahide Garda Station -6.152084  53.451088
26           Mountjoy Garda Station -6.266834   53.36053
27      Pearse Street Garda Station -6.256193  53.345689
28             Raheny Garda Station -6.178056  53.378896
29          Rathcoole Garda Station -6.464697  53.281814
30        Rathfarnham Garda Station -6.288549   53.29741
31          Rathmines Garda Station -6.266966  53.321368
32         Ronanstown Garda Station -6.406166  53.337878
33             Santry Garda Station -6.250638  53.389967
34           Shankill Garda Station -6.120737  53.233488
35           Skerries Garda Station -6.106979  53.579586
36       Store Street Garda Station -6.251706  53.350485
37      Sundrive Road Garda Station -6.298697    53.3302
38             Swords Garda Station -6.221112  53.456024
39           Tallaght Garda Station -6.367589  53.286807
40           Terenure Garda Station  -6.28781   53.30999

In [23]:
df.to_csv('garda_coords.csv')